### PR TITLE
Standalone vault

### PR DESCRIPTION
**Pull request by jenow**
## What does this PR do?

Make Vault standalone.

Please read the [README](https://github.com/kuzzleio/kuzzle-vault/blob/4f89fef70c680f83654757120ea7bf872eedb695/README.md) to know more.

### How should this be manually tested?

```sh
npm run test
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+```js
+encrypt([outputFile, replaceFileIfExist]);
+```
+
+**Arguments**
+
+| Name | Type              | Description |
+| -------- | ----------------- | ----------- |
+| `outputFile`  | <pre>string</pre> | Optional file used to store the encrypted secrets. If not set `encryptedSecretsFile` from the [constructor](#constructor) will be used instead |
+| `replaceFileIfExist`  | <pre>bool</pre> | Optional argument to overwrite the file if it already exist |
```
Comment (y-abs) [inline README.md]:
```suggestion
| `replaceFileIfExist`  | <pre>bool</pre> | Optional argument to overwrite the file if it already exists |
```

Also, should be `replaceFileIfExists`

File: lib/vault.js [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    if (!outputFile && !this.encryptedSecretsFile) {
+      const error = new Error('No outputFile or encryptedSecretsFile specified.');
+
+      error.name = 'PreconditionError';
+      throw error;
+    }
+
+    const fileExists = fs.existsSync(this.encryptedSecretsFile);
+    if (fileExists && !replaceFileIfExist) {
+      const error = new Error('Ouput file already exists.');
```
Comment (y-abs) [inline lib/vault.js]:
```suggestion
      const error = new Error('Output file already exists.');
```

File: lib/vault.js [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+      throw error;
+    }
+
+    const secrets = JSON.parse(fs.readFileSync(this.secretsFile, 'utf-8'));
+    const encryptedSecrets = this._encryptObject(secrets);
+    fs.writeFileSync(outputFile || this.encryptedSecretsFile, JSON.stringify(encryptedSecrets, null, 2));
+  }
+
+  /**
+   * Iterate recursively through object values and try to
```
Comment (y-abs) [inline lib/vault.js]:
```suggestion
   * Iterates recursively through object values and tries to
```

File: lib/vault.js [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+      } else if (_.isString(value)) {
+        secrets[key] = this._decryptData(value);
+      }
+    }
+
+    return secrets;
+  }
+
+  /**
+   * Iterate recursively through object values and try to
```
Comment (y-abs) [inline lib/vault.js]:
```suggestion
   * Iterates recursively through object values and tries to
```

File: lib/vault.js [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+      } else if (_.isString(value)) {
+        encryptedSecrets[key] = this._encryptData(value);
+      }
+    }
+
+    return encryptedSecrets;
+  }
+
+  /**
+   * Encrypt data with AES CBC using the secret key and an initialization vector
```
Comment (y-abs) [inline lib/vault.js]:
```suggestion
   * Encrypts data with AES CBC using the secret key and an initialization vector
```

File: lib/vault.js [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+      iv = crypto.randomBytes(16),
+      cipher = crypto.createCipheriv('aes-256-cbc', this.vaultKeyHash, iv);
+
+    const encryptedData = cipher.update(data, 'utf8', 'hex') + cipher.final('hex');
+
+    return `${encryptedData}.${iv.toString('hex')}`;
+  }
+
+  /**
+   * Decrypt data with AES CBC using the secret key and the initialization vector
```
Comment (y-abs) [inline lib/vault.js]:
```suggestion
   * Decrypts data with AES CBC using the secret key and the initialization vector
```

File: lib/vault.js [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+   * This function keep compatibility with old IV size (8 bytes) from 1.8.0 to 1.8.1
```
Comment (y-abs) [inline lib/vault.js]:
```suggestion
   * This function keeps compatibility with old IV size (8 bytes) from 1.8.0 to 1.8.1
```

**Review by y-abs — CHANGES_REQUESTED** [review]

File: lib/vault.js [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+   *
+   * @param {string|null} vaultKey - key used to decrypt the secrets
+   */
+  constructor (vaultKey = null, secretsFile = null, encryptedSecretsFile = null) {
+    this.secrets = {};
+    this.secretsFile = secretsFile;
+    this.encryptedSecretsFile = encryptedSecretsFile;
+
+    if (_.isString(vaultKey) && vaultKey.length > 0) {
+      this.vaultKey = vaultKey;
```
Comment (scottinet) [inline lib/vault.js]: It looks like this property can be set after the object has been instantiated. In that case, shouldn't we add a setter for it, computing this.vaultKeyHash whenever a vaultKey is set?

File: lib/vault.js [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+   *  - a key is found but no file
+   *  - a file is found but no key
+   *
+   * @param {string|null} secretsFile - file containing the encrypted secrets
+   */
+  encrypt (outputFile = null, replaceFileIfExist = false) {
+    if (!outputFile && !this.encryptedSecretsFile) {
+      const error = new Error('No outputFile or encryptedSecretsFile specified.');
+
+      error.name = 'PreconditionError';
```
Comment (scottinet) [inline lib/vault.js:67]: Why do we want to (incompletely) mimic a KuzzleError object here? Can't we just throw a normal error, and let Kuzzle decide what kind of error it should forward to users? As for the CLI, I don't think it needs KuzzleError instances.

File: lib/vault.js [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+    if (!this.vaultKey) {
+      const error = new Error('Vault key not found.');
+      
+      error.name = 'PreconditionError';
+      throw error;
+    }
+
+    const secrets = {};
+
+    for (const key of Object.keys(encryptedSecrets)) {
```
Comment (scottinet) [inline lib/vault.js]: (nitpicking) kuzzle v2 == node 10 == nifty features :tada: 

```js
    for (const [key, value] of Object.entries(encryptedSecrets)) {
```

File: package.json [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    "url": "git+https://github.com/kuzzleio/kuzzle-vault.git"
+  },
+  "author": "The Kuzzle Team <support@kuzzle.io>",
+  "license": "Apache-2.0",
+  "bugs": {
+    "url": "https://github.com/kuzzleio/kuzzle-vault/issues"
+  },
+  "homepage": "https://github.com/kuzzleio/kuzzle-vault#readme",
+  "dependencies": {
+    "assert": "^2.0.0",
```
Comment (scottinet) [inline package.json]: I don't see that module used. Furthermore, this module aims to provide [Node.js assert](https://nodejs.org/dist/latest-v10.x/docs/api/assert.html) to browsers, so I don't think we need it.

File: package.json [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    "crypto": "^1.0.1",
```
Comment (scottinet) [inline package.json]: This package is deprecated. Use [Node.js crypto](https://nodejs.org/dist/latest-v10.x/docs/api/crypto.html) instead.

**Review by scottinet — CHANGES_REQUESTED** [review]

File: lib/vault.js (hunk 9 above)
Comment (jenow) [reply lib/vault.js:67]: I do this so Kuzzle can identify what kind of error this is and avoid doing a
```js
if (error.message === 'No outputFile or encryptedSecretsFile specified.') {
```

**Review by scottinet — APPROVED** [review]

**Comment by jenow** [comment]
We decided to make the Vault standalone (https://github.com/kuzzleio/kuzzle-vault) therefor it is now useless to have an API endpoint to encrypt/decrypt secrets.

**Review by Aschen — APPROVED** [review]

**Pull request merged**